---
title: "Ampliseq WECARE NFFE PCA"
author: "Alexey Larionov"
date: "29 Dec 2021"
output:
  html_document:
    toc: true
    number_sections: true
    toc_float: true
editor_options:
  chunk_output_type: console
---

# Summary

Had to suppress the smoothing log-p-values: see details in the script  

PCA for ampliseq-NFFE (b37) dataset w/o Danish:  

- Shows no irregularities in PCA analysis  
- Provides evidence for using top 2 PC-s in regression model  
- There was possibly one PCA outlier; however, with such small number of variants it was likely to be spurious; also it was in the 5th PC (at the best), so it was not excluded   

<style>
pre{
  overflow-x: auto;
}
pre code{
  word-wrap: normal;
  white-space: pre;
}
</style>

```{r echo=F}
options(width=999)
```

# Start section

```{r}

# Time stamp
Sys.time()

# Clenan-up
rm(list=ls())
graphics.off()

# Memory
gc()

# Options
options(stringsAsFactors = F,
        warnPartialMatchArgs = T, 
        warnPartialMatchAttr = T, 
        warnPartialMatchDollar = T)

# Working folders
base_folder="/Users/alexey/Documents/wecare/final_analysis_2021/reanalysis_afterNov2021/ampliseq/s02_ampliseq_wecare_nfe"
data_folder <- file.path(base_folder,"data","s07_relatedness_and_pca")
scripts_folder <- file.path(base_folder,"scripts","s07_relatedness_and_pca")
setwd(scripts_folder)

# Libraries
library(bigsnpr) # for bed_autoSVD() and bed()
library(bigutilsr) # for prob_dist() and tukey_mc_up() for outlier detection
library(hexbin) # for plotting svd loadings
library(ggplot2)
library(dplyr)

NCORES <- 1
#NCORES <- nb_cores() # 2

```

# Read bed-bim-fam file-set

```{r}

# Location of bed file
bed_file <- file.path(data_folder,"s02_bed_bim_fam","common_biallelic_autosomal_snps_in_HWE.bed")

# Attach PLINK data to R environment
wecare.bed <- bed(bed_file) # bigsnpr::bed

# Explore wecare.bed
wecare.bed
names(wecare.bed)
#attributes(wecare.bed)
#str(wecare.bed)
#wecare.bed$bedfile
#wecare.bed$address

# Clean-up
rm(bed_file)

```

# Phenotypes

```{r}

# Phenotypes from plink
wecare_fam.df <- wecare.bed$fam
dim(wecare_fam.df)
head(wecare_fam.df)
table(wecare_fam.df$affection, useNA = "always")

# Phenotypes from the main dataset file
load(file.path(base_folder,"data","s06_qc_filters","s04_filter_samples_by_call_rate.RData"))
rm(genotypes.mx,variants.df)

# Update folders
base_folder="/Users/alexey/Documents/wecare/final_analysis_2021/reanalysis_afterNov2021/ampliseq/s02_ampliseq_wecare_nfe"
data_folder <- file.path(base_folder,"data","s07_relatedness_and_pca")
scripts_folder <- file.path(base_folder,"scripts","s07_relatedness_and_pca")

dim(phenotypes.df)
colnames(phenotypes.df)
table(phenotypes.df$cc, useNA = "always")

#str(phenotypes.df)
#table(phenotypes.df$filter)
#phenotypes.df[phenotypes.df$filter=="eigenvectors_outlier","wes_id"]
#"Possibly_related_to_P1-D05" -> phenotypes.df[phenotypes.df$wes_id=="P5_C12","filter"]

# Merge fam-file and phenotypes from the main dataset 
# (removing samples that are not in fam-file) 
wecare_phenotypes.df <- left_join(wecare_fam.df, phenotypes.df,
                                  by=c("sample.ID"="long_ids"))
dim(wecare_phenotypes.df)
colnames(wecare_phenotypes.df)
#sum(wecare_phenotypes.df$sample.ID == "P5_C12")

# Make sure that dplyr::left_joint hasnt changed the order of samples
sum(wecare_phenotypes.df$sample.ID != wecare_fam.df$sample.ID)

# Add column for outliers
wecare_phenotypes.df <- data.frame(wecare_phenotypes.df,outlier=F)

# Clean-up
rm(phenotypes.df, wecare_fam.df)

```

# Explore variants

Data from bed-bim-fam only: explored; not needed for PCA calculation

```{r}

# map file
wecare_map.df <- wecare.bed$map
dim(wecare_map.df)
head(wecare_map.df)

# make simple counts 
wecare_maf.df <- bed_MAF(wecare.bed)
dim(wecare_maf.df)
head(wecare_maf.df)

# merge map file with the counts
wecare_variants.df <- cbind(wecare_map.df,wecare_maf.df)
dim(wecare_variants.df)
head(wecare_variants.df)

# Variants with AF(ref) < AF(alt)
inverted <- wecare_variants.df$ac != wecare_variants.df$mac
sum(inverted)
#wecare_variants.df[inverted,]

# Clean-up
rm(wecare_map.df, wecare_maf.df, inverted)

```

# PCA

Only one round was performed.  
However, it could be sequentially repeated if outliers are detected 

Takes care about LD etc.  
See ?plot.big_SVD for plotting svd objets.  

```{r}

# Get indices of non-outliers in format required by bed_autoSVD
# (integer indices, indicating row numbers)
table(wecare_phenotypes.df$outlier, useNA = "always")
non_outliers1 <- which(!wecare_phenotypes.df$outlier)
length(non_outliers1)

# bigsnpr::bed_autoSVD, Default k = 10
#using non-outlier samples (ind.row) and all variants (ind.col) 
#table(wecare.bed$map$chromosome) - if complains abotut non-numeric chromosomes
wecare.svd1 <- bed_autoSVD(wecare.bed, 
                          ind.row=non_outliers1, 
                          ncores = NCORES,
                          roll.size=0) 

```

Because of the low number of variants, 
I had to suppress smoothing log-p-values
to pass this point.  

Other way it crashed at this point with the following error message:  

Phase of clumping (on MAC) at r^2 > 0.2.. keep 152 variants.
Discarding 0 variant with MAC < 10.

Iteration 1:
Computing SVD..
Error in bigutilsr::rollmean(S.col[ind], roll.size) : 
  Parameter 'size' is too large.

This bug/question was not yet resolved as on 29May2021:  
https://github.com/privefl/bigutilsr/issues/2  

An alternative for suppressing the smoothing of log-p-values would be 
a simplifyed procedure with PLINK for PCA instead of using bigsnpr.  

```{r}
#ind.col=vars_not_in_LD, 

# Variants not in LD (detected by clumping during autoSVD)
vars_not_in_LD1 <- attr(wecare.svd1, "subset")
length(vars_not_in_LD1)

#attributes(wecare.svd)
str(wecare.svd1)

# Eigenvalues
length(wecare.svd1$d)
wecare.svd1$d
plot(wecare.svd1) # default type="screeplot" see ?plot.big_SVD  

# Eigenvectors
dim(wecare.svd1$u)
head(wecare.svd1$u)

# PCA summary (for PCs from 1 to 10)
plot(wecare.svd1,type = "scores",scores=1:10,coeff=0.4)

# Loadings
dim(wecare.svd1$v)
head(wecare.svd1$v)

# Loadings summary (for PCs from 1 to 10)
plot(wecare.svd1,type="loadings",loadings=1:10,coeff=0.4)

# Calculate a measure of "outlieness"  
U1 <- wecare.svd1$u
prob1 <- prob_dist(U1, ncores=NCORES) # bigutilsr::prob_dist
S1 <- prob1$dist.self / sqrt(prob1$dist.nn)
tukey_threshold1 <- tukey_mc_up(S1) # bigutilsr::tukey_mc_up

# Outliers
outliers1 <- S1 >= tukey_threshold1
sum(outliers1)
 
# Histogram by "outlieness" score
ggplot() +
  geom_histogram(aes(S1), color = "black", fill = "blue", alpha = 0.3) +
  theme_bigstatsr() +
  geom_vline(xintercept=tukey_threshold1, colour="red") +
  labs(x = "Statistic of outlierness (S)", y = "Frequency (sqrt-scale)")

```

# Update phenotypes table

```{r}

phenotypes_with_PCs.df <- wecare_phenotypes.df
eigenvectors.mx <- wecare.svd1$u

#phenotypes_with_PCs.df <- wecare_phenotypes.df[!wecare_phenotypes.df$outlier,]
#eigenvectors.mx <- wecare.svd2$u

dim(phenotypes_with_PCs.df)
dim(eigenvectors.mx)

colnames(eigenvectors.mx) <- 
  c("pc1","pc2","pc3","pc4","pc5","pc6","pc7","pc8","pc9","pc10")

phenotypes_with_PCs.df <- cbind(phenotypes_with_PCs.df, eigenvectors.mx)
dim(phenotypes_with_PCs.df)
colnames(phenotypes_with_PCs.df)

# Clean-up
rm(wecare_phenotypes.df, eigenvectors.mx) # vars_not_in_LD1
#somehow close wecare.bed ?

```

# Compare with PC-s calculated by PLINK

```{r}

plink_pcs_file <- file.path(data_folder,"s05_plink_pca","ampliseq_only_20PCs.eigenvec")
plink_pcs.df <- read.table(plink_pcs_file, header=T, sep="\t")
rownames(plink_pcs.df) <- plink_pcs.df$IID
plink_pcs.df <- plink_pcs.df[phenotypes_with_PCs.df$sample.ID,]
dim(plink_pcs.df)

# Check consistency with the previous eigenvectors from WES
plot(plink_pcs.df$PC1,
     phenotypes_with_PCs.df$pc1,
     main="PC1: PLINK vs bed_autoSVD")

plot(plink_pcs.df$PC2,
     phenotypes_with_PCs.df$pc2,
     main="PC2: PLINK vs bed_autoSVD")

rm(plink_pcs_file, plink_pcs.df)

```

# Calculate PC-outliers by 6xSD

```{r}

sd_threshold <- 6

# PC1 outliers
pc1 <- phenotypes_with_PCs.df$pc1
pc1_mean <- mean(pc1)
pc1_sd  <- sd(pc1)
lo_pc1 <- pc1 < pc1_mean - sd_threshold * pc1_sd
hi_pc1 <- pc1 > pc1_mean + sd_threshold * pc1_sd

cat("pc1 lo/hi:",sum(lo_pc1),"/",sum(hi_pc1),"\n")

phenotypes_with_PCs.df$sample.ID[lo_pc1]
phenotypes_with_PCs.df$sample.ID[hi_pc1]

# PC2 outliers
pc2 <- phenotypes_with_PCs.df$pc2
pc2_mean <- mean(pc2)
pc2_sd  <- sd(pc2)
lo_pc2 <- pc2 < pc2_mean - sd_threshold * pc2_sd
hi_pc2 <- pc2 > pc2_mean + sd_threshold * pc2_sd

cat("pc2 lo/hi:",sum(lo_pc2),"/",sum(hi_pc2),"\n")

phenotypes_with_PCs.df$sample.ID[lo_pc2]
phenotypes_with_PCs.df$sample.ID[hi_pc2]

rm(sd_threshold,
   pc1, pc1_mean, pc1_sd, lo_pc1, hi_pc1,
   pc2, pc2_mean, pc2_sd, lo_pc2, hi_pc2)

```

# Detailed PCA plots

```{r}

selected_svd <- wecare.svd1
#selected_svd <- wecare.svd2

plot(selected_svd, type = "scores") +
     aes(color = as.factor(phenotypes_with_PCs.df$cc)) +
     labs(title = NULL, color = "Case") 

# + geom_hline(pc2_mean - sd_threshold * pc2_sd,
#              linetype="dashed", color = "red")

plot(selected_svd, type = "scores", scores=3:4) +
     aes(color = as.factor(phenotypes_with_PCs.df$cc)) +
     labs(title = NULL, color = "Case")

plot(selected_svd, type = "scores", scores=5:6) +
     aes(color = phenotypes_with_PCs.df$affection == 2) +
     labs(title = NULL, color = "Case")

plot(selected_svd, type = "scores", scores=7:8) +
     aes(color = as.factor(phenotypes_with_PCs.df$cc)) +
     labs(title = NULL, color = "Case")

plot(selected_svd, type = "scores", scores=9:10) +
     aes(color = as.factor(phenotypes_with_PCs.df$cc)) +
     labs(title = NULL, color = "Case")

rm(selected_svd)

```

# Save results

```{r}
rm(NCORES,wecare.svd1,wecare.bed,wecare_variants.df)
save.image(file.path(data_folder,"s06_calculate_PCs.RData"))
save(phenotypes_with_PCs.df,file=file.path(data_folder,"s06_phenotypes_with_PCs.RData"))

```

# End section

```{r}

ls()
Sys.time()
gc()

```
