---
title: "Ampliseq: BC9 SKAT"
author: "Alexey Larionov"
date: "20 Apr 2022"
output: 
  html_document:
    toc: true
    toc_float: true
editor_options: 
  chunk_output_type: console
---

# Summary

SKAT Burden test for functionally important variants in BC9 genes.  

The model includes:  

- 2 top PCs  
- Age at the 1st event & time to the 2nd event  
- Num of pregnancies (coded 1-2-3)  
- All treatments: hormone, chemo, and **br_xray**  

Imputing one case with missed data on hormonal treatment.  

Calculate model using default imputation & weights; and using efficient resampling.  

Variants selection (done in the previous script): 

*Known or predicted pathogenic*  
**Known pathogenic**  
- Annotated as pathogenic/likely pathogenic in ClinVar (v?)  
**Predicted pathogenic**  
- LoF (HIGH importance by VEP)  
- Consensus high grade missenses: deleterious by SIFT & probably_damaging by PolyPhen & CADD >=25  

*Rare*  
- Rare (AF in Wecare < 0.01)  

*Exclude known benign*  
- Excluding known benign in ClinVar  

<style>
pre{
  overflow-x: auto;
}
pre code{
  word-wrap: normal;
  white-space: pre;
}
</style>

```{r echo=F}
options(width=999)
```

# Start section

```{r}

# Time
Sys.time()

# Memory
gc()

# Clean up
rm(list=ls())
graphics.off()

# Options
options(stringsAsFactors = F)

# Working folders
base_folder="/Users/alexey/Documents/wecare/final_analysis_2021/reanalysis_afterNov2021/ampliseq/s01_ampliseq_wecare_only"
data_folder <- file.path(base_folder,"data","s09_skat")
#dir.create(data_folder)
scripts_folder <- file.path(base_folder,"scripts","s09_skat")
setwd(scripts_folder)

# Libraries
library(SKAT)
library(dplyr)

```

# Load data

```{r}

load(file.path(base_folder,"data","s08_crude_counts","s03_allele_counts_bc9.RData"))

# Update folders
base_folder="/Users/alexey/Documents/wecare/final_analysis_2021/reanalysis_afterNov2021/ampliseq/s01_ampliseq_wecare_only"
data_folder <- file.path(base_folder,"data","s09_skat")
scripts_folder <- file.path(base_folder,"scripts","s09_skat")

```

# Check data

```{r}

ls()

dim(genotypes.mx)
dim(variants.df)
dim(phenotypes.df)
table(phenotypes.df$cc, useNA = "always")

# Check consistency of rownames and colnames
sum(colnames(genotypes.mx) != phenotypes.df$sample.ID)
sum(rownames(genotypes.mx) != variants.df$ID)

dim(bc9_variants.df)
dim(bc9_fiv.df)

```

# Explore phenotypes

## UBC-CBC counts

```{r}

table(phenotypes.df$cc, useNA = "always")

ubc <- phenotypes.df$cc == 0
cbc <- phenotypes.df$cc == 1

```

## Age at 1st cancer

```{r}

mean(phenotypes.df$sub_dx_age[ubc])
sd(phenotypes.df$sub_dx_age[ubc])

mean(phenotypes.df$sub_dx_age[cbc])
sd(phenotypes.df$sub_dx_age[cbc])

t.test(phenotypes.df$sub_dx_age[ubc], phenotypes.df$sub_dx_age[cbc])

```

## Ref time

```{r}

mean(phenotypes.df$rstime[ubc])
sd(phenotypes.df$rstime[ubc])

mean(phenotypes.df$rstime[cbc])
sd(phenotypes.df$rstime[cbc])

t.test(phenotypes.df$rstime[ubc], phenotypes.df$rstime[cbc])

```

## Stage

```{r}

table(phenotypes.df[,c("stage_fd","cc")], useNA = "always")
x <- table(phenotypes.df[,c("stage_fd","cc")])
x
fisher.test(x)
rm(x)

```

## ER

```{r}

x <- table(phenotypes.df[,c("er1_cat","cc")], useNA = "ifany")

x

chisq.test(x)
rm(x)

```

## Cytotoxic chemo

```{r}

table(phenotypes.df[,c("chemo_self_mra","cc")], useNA = "always")

x <- table(phenotypes.df[,c("chemo_self_mra","cc")])

x

fisher.test(x)
rm(x)

```

## Hormonal therapy

```{r}

x <- table(phenotypes.df[,c("hormone_self_mra","cc")], useNA = "ifany")

rownames(x) <- c("no","yes")
x

fisher.test(x)
rm(x)

```

## Breast x-ray

```{r}

table(phenotypes.df[,c("xrtbreast","cc")], useNA = "always")

x <- table(phenotypes.df[,c("xrtbreast","cc")])
x
fisher.test(x)
rm(x)

```

## Num pregn

```{r}

table(phenotypes.df$Num_ftp_fd, useNA = "always")
num_preg <- as.integer(phenotypes.df$Num_ftp_fd)

0 -> num_preg[as.integer(phenotypes.df$Num_ftp_fd) <= 0] # -1,0
1 -> num_preg[as.integer(phenotypes.df$Num_ftp_fd) %in% c(1,2)] # 1,2
2 -> num_preg[as.integer(phenotypes.df$Num_ftp_fd) >= 3] # 3+

table(num_preg)

table(num_preg,phenotypes.df$cc, useNA = "always")

x <- table(num_preg,phenotypes.df$cc)
rownames(x) <- c("None","1-2","3+")
x

chisq.test(x)
rm(x)

```

## Family history

```{r}

table(phenotypes.df[,c("family_history","cc")], useNA = "always")

x <- table(phenotypes.df[,c("family_history","cc")])
x
x <- x[c(3,1,2),]
x
chisq.test(x)
rm(x, cbc, ubc)

```

# Prepare data for model

No missed data in outcomes, covariates or eigenvectors.  

Missed genotypes are imputed by SKAT as average AF  
(default impute.method="fixed" uses mean genotype values).  

## Outcomes

```{r}

Y <- phenotypes.df$cc
length(Y)
sum(is.na(Y)) # 0

```

## Genotypes

Select variants and transpose

```{r}

G <- t(genotypes.mx[bc9_fiv.df$ID,])
dim(G)
sum(is.na(G))

# Do NOT impute missed genotypes (they are handled by SKAT)
sum(is.na(G))/(ncol(G)*nrow(G)) # ~1.7%

```

## Eigenvectors

```{r}

E <- as.matrix(phenotypes.df[,c("pc1","pc2")])
dim(E)
sum(is.na(E)) # 0

```

## "Matching" covariates 

Originally, the cases and controls were matched by Age at diagnosis and 
time to event / lack of event ("age_dx", "rstime"). 
Hence the "M" abbreviation (called SKAT-M in preliminary analyses). 

At the same time, SKAT does not allow using this matching information.  
Also, some samples failed sequencing, thus breaking the pairs. 
Overall, the compared groups are very well balanced by "age_dx", "rstime". 
Still, DC recommended to use these covariates to account for differences at 
individual level.  

By including these we assume linear effect of these confaunders.  
This is supported by the previous biological knowledge. 

```{r}

M <- as.matrix(phenotypes.df[,c("sub_dx_age", "rstime")])
sum(is.na(M)) # 0

```

## Treatment

Assuming that 0 means no treatment

```{r}

table(phenotypes.df$hormone_self_mra, useNA = "always")
hormone <- phenotypes.df$hormone_self_mra
table(hormone, useNA = "always")

table(phenotypes.df$chemo_self_mra, useNA = "always")
chemo <- phenotypes.df$chemo_self_mra
table(chemo, useNA = "always")

table(phenotypes.df$xrtbreast, useNA = "always")
xray <- phenotypes.df$xrtbreast
table(xray, useNA = "always")

```

## Number of pregnancies

Code as 0,1,2 (it should have been done earlier during the data clean-up)  

```{r}

table(phenotypes.df$Num_ftp_fd, useNA = "always")
num_preg <- as.integer(phenotypes.df$Num_ftp_fd)

0 -> num_preg[as.integer(phenotypes.df$Num_ftp_fd) <= 0] # -1,0
1 -> num_preg[as.integer(phenotypes.df$Num_ftp_fd) %in% c(1,2)] # 1,2
2 -> num_preg[as.integer(phenotypes.df$Num_ftp_fd) >= 3] # 3+

table(num_preg)

```

## Combined "Clinical" covariates

```{r}

C <- cbind(chemo, hormone, xray, num_preg) 
sum(is.na(C)) # 0

# Clean-up
rm(chemo, hormone, xray, num_preg)

```

# SKAT

## Null

D for the dichotomous outcome

```{r}

skat_null <- SKAT_Null_Model(Y ~ C + E + M, out_type="D", n.Resampling = 10000)

```

## Burden 

Calculate burden test with  

- efficient resampling  
- default weights beta[1,25]  

```{r}

burden_test <- SKATBinary(G, skat_null, method="Burden", method.bin = "ER")
burden_test$p.value
burden_test$MAC
names(burden_test)

# Clean-up
rm(Y, G, E, C, M, skat_null)

```

# Data summary

```{r}

ls()

dim(genotypes.mx)
dim(variants.df)
dim(phenotypes.df)
table(phenotypes.df$cc, useNA = "always")

# Check consistency of rownames and colnames
sum(colnames(genotypes.mx) != phenotypes.df$sample.ID)
sum(rownames(genotypes.mx) != variants.df$ID)

dim(bc9_variants.df)
dim(bc9_fiv.df)

```

# Save results

```{r}

save.image(file.path(data_folder,"s01_skat_bc9.RData"))

```

# Final section

```{r}

ls()
sessionInfo()
Sys.time()

```
