---
title: "Ampliseq wecare NFFE: filter genotypes by dp and gq"
author: "Alexey Larionov"
date: "27 Dec 2021"
output:
  html_document:
    toc: true
    number_sections: true
    toc_float: true
editor_options:
  chunk_output_type: console
---

# Summary

Remove (set to NA) ~ 17% genotypes.  

Genotypes removed if any of these conditions is true:  

- gq < 20  : ~11% genotypes
- dp < 10  : ~5% genotypes more
- dp > 859  : ~1% genotypes more, 10x mean dp (was 360 in WES)

Then remove non-polymorphic sites, if appear after the genotypes filtering.  

gq 20 filter is set arbitrary; however, it is consistent with what some previous papers (e.g. see Carson BMC Bioinformatics. 2014 15:125).  

Check genotypes (missingness, dp and gq) and call rates (per variant and per sample) before and after filtering.  

Input data: 7,322 variants x 587 samples (198 NFFE, 194 UBC & 195 CBC)  
Output data: 6,194 variants x 587 samples (198 NFFE, 194 UBC & 195 CBC)  

<style>
pre{
  overflow-x: auto;
}
pre code{
  word-wrap: normal;
  white-space: pre;
}
</style>

```{r echo=F}
options(width=999)
```

# Start section

```{r}

# Time
Sys.time()

# Memory
gc()

# Clean up
rm(list=ls())
graphics.off()

# Options
options(stringsAsFactors = F)

# Working folders
base_folder="/Users/alexey/Documents/wecare/final_analysis_2021/reanalysis_afterNov2021/ampliseq/s02_ampliseq_wecare_nfe"
data_folder <- file.path(base_folder,"data","s06_qc_filters")
#dir.create(data_folder)
scripts_folder <- file.path(base_folder,"scripts","s06_qc_filters")
setwd(scripts_folder)

# Thresholds for genotypes
min.gq <- 20
min.dp <- 10
max.dp <- 1100 # 10x mean dp

```

# Load data

```{r}

load(file.path(base_folder,"data","s05_update and_cleaning","s05_update_variant_annotations_apply_FILTER.RData"))

# Update folders
base_folder="/Users/alexey/Documents/wecare/final_analysis_2021/reanalysis_afterNov2021/ampliseq/s02_ampliseq_wecare_nfe"
data_folder <- file.path(base_folder,"data","s06_qc_filters")
scripts_folder <- file.path(base_folder,"scripts","s06_qc_filters")

```

# Check data

```{r}

# List objects
ls()

# Check sizes
dim(gt_add.mx)
dim(gt_num.mx)
dim(gt_chr.mx)

dim(alt_fraction.mx)

dim(gq.mx)
dim(dp.mx)

dim(variants.df)
dim(phenotypes.df)
table(phenotypes.df$cc, useNA = "always")

# Check consistency of rownames and colnames

sum(rownames(gt_add.mx) != rownames(gt_num.mx))
sum(rownames(gt_add.mx) != rownames(gt_chr.mx))
sum(rownames(gt_add.mx) != rownames(alt_fraction.mx))
sum(rownames(gt_add.mx) != rownames(gq.mx))
sum(rownames(gt_add.mx) != rownames(dp.mx))

sum(colnames(gt_add.mx) != colnames(gt_num.mx))
sum(colnames(gt_add.mx) != colnames(gt_chr.mx))
sum(colnames(gt_add.mx) != colnames(alt_fraction.mx))
sum(colnames(gt_add.mx) != colnames(gq.mx))
sum(colnames(gt_add.mx) != colnames(dp.mx))

sum(rownames(gt_add.mx) != rownames(variants.df))
sum(colnames(gt_add.mx) != rownames(phenotypes.df)) 

```

# Explore genotypes before filtering

- gt missingness  
- gq and dp in non-NA genotypes  

```{r}

# Fraction of NA genotypes
sum(is.na(gt_add.mx))/(nrow(gt_add.mx)*ncol(gt_add.mx))
gt_add.mx[1:5,1:5]

# Index of non-NA gt  
non_NA_gt <- !is.na(gt_add.mx)

# Mean GQ and DP in non-NA genotypes
quantile(gq.mx[non_NA_gt], na.rm=TRUE)
quantile(dp.mx[non_NA_gt], na.rm=TRUE)

mean(gq.mx[non_NA_gt], na.rm=TRUE)
mean(dp.mx[non_NA_gt], na.rm=TRUE)

# gq  (when gt is not NA !)
hist(gq.mx[non_NA_gt], 
     main="Histogram of gq in non-NA genotypes", xlab="gq")
abline(v=min.gq, lty=2, col="red")

# dp (when gt is not NA !)
hist(dp.mx[non_NA_gt], 
     main="Histogram of dp in non-NA genotypes", xlab="dp")
abline(v=max.dp, lty=2, col="red")

hist(dp.mx[non_NA_gt & dp.mx < 100], 
     main="Histogram of dp in non-NA genotypes (zoom to <100)", xlab="dp")
abline(v=min.dp, lty=2, col="red")

# crude estimates for proportions of genotypes to be removed by filters
sum(gq.mx < min.gq, na.rm = T) / sum(non_NA_gt)
sum(dp.mx < min.dp, na.rm = T) / sum(non_NA_gt)
sum(dp.mx > max.dp, na.rm = T) / sum(non_NA_gt)

# Clean-up
rm(non_NA_gt)

```

# Variant call rates before filtering

```{r}

# Function to calculate call rate
call_rate.udf <- function(x){1-sum(is.na(x))/length(x)}

# Call rates per variant
var_call_rate <- apply(gt_add.mx,1,call_rate.udf)

hist(var_call_rate, breaks=50, main="Call rate per variant (before filters)", xlab="Call rates")
abline(v=0.85, lty=2, col="red")

plot(var_call_rate, main="Call rate per variant (before filters)")
abline(h=0.85, lty=2, col="red")

# SKAT requires variaint call rate > 85%
sum(var_call_rate < 0.85)/length(var_call_rate)

# Clean-up
rm(var_call_rate)

```

# Sample call rates before filtering

```{r}

# Call rates per sample
case_call_rate <- apply(gt_add.mx,2,call_rate.udf)

hist(case_call_rate, xlim=c(0,1), main="Call rate per sample (before filters)", xlab="Call rates")
abline(v=0.5, lty=2, col="red")

plot(case_call_rate, ylim=c(0,1), main="Call rate per sample (before filters)")
abline(h=0.5, lty=2, col="red")

# Clean-up
rm(case_call_rate)

```

# Remove low gq

```{r}

# num and fraction of genotypes to be removed (in !NA gt)
non_NA_gt <- !is.na(gt_add.mx)
sum(gq.mx[non_NA_gt] < min.gq, na.rm=TRUE)/sum(non_NA_gt)

# Apply filter (to gt only !)
NA -> gt_add.mx[ gq.mx < min.gq ]
NA -> gt_num.mx[ gq.mx < min.gq ]
NA -> gt_chr.mx[ gq.mx < min.gq ]

# Clean up
rm(non_NA_gt)

```

# Remove low dp

```{r}

# num and fraction of genotypes to be removed (in !NA gt after gq filtering)
non_NA_gt <- !is.na(gt_add.mx)
sum(dp.mx[non_NA_gt] < min.dp, na.rm=TRUE)/sum(non_NA_gt)

# Apply filter (to gt only !)
NA -> gt_add.mx[ dp.mx < min.dp ]
NA -> gt_num.mx[ dp.mx < min.dp ]
NA -> gt_chr.mx[ dp.mx < min.dp ]

# Clean up
rm(non_NA_gt)

```

# Remove (too) high dp

```{r}

# num and fraction of genotypes to be removed (in !NA gt after gq filtering)
non_NA_gt <- !is.na(gt_add.mx)
sum(dp.mx[non_NA_gt] > max.dp, na.rm=TRUE)/sum(non_NA_gt)

# Apply filter (to gt only !)
NA -> gt_add.mx[ dp.mx > max.dp ]
NA -> gt_num.mx[ dp.mx > max.dp ]
NA -> gt_chr.mx[ dp.mx > max.dp ]

# Clean up
rm(non_NA_gt)

```

# Remove non-polymorphic sites

After genotypes filtering some variant sites may become non-polymorphic

```{r}

# Function to detect uniform numeric vector
uniformity_check.udf <- function(x){
  if (all(is.na(x))){"All_NA"}
  else if (min(x,na.rm=T)==max(x,na.rm=T)){"Uniform"}
  else {"Non-Uniform"}}
    
uniformity_check <- apply(gt_add.mx,1,uniformity_check.udf)
summary(as.factor(uniformity_check))
non_uniform_sites <- uniformity_check == "Non-Uniform"

# Update matrices and data-frame(s)
gt_add.mx <- gt_add.mx[non_uniform_sites,]
gt_num.mx <- gt_num.mx[non_uniform_sites,]
gt_chr.mx <- gt_chr.mx[non_uniform_sites,]

gq.mx <- gq.mx[non_uniform_sites,]
dp.mx <- dp.mx[non_uniform_sites,]

alt_fraction.mx <- alt_fraction.mx[non_uniform_sites,]

variants.df <- variants.df[non_uniform_sites,]

# Clean-up
rm(uniformity_check.udf,uniformity_check,non_uniform_sites)

```

# Explore genotypes after filtering

- gt missingness  
- gq and dp in non-NA genotypes  

```{r}

# Fraction of NA genotypes
sum(is.na(gt_add.mx))/(nrow(gt_add.mx)*ncol(gt_add.mx))
gt_add.mx[1:5,1:5]

# Index of non-NA gt  
non_NA_gt <- !is.na(gt_add.mx)

# gq  (when gt is not NA !)
hist(gq.mx[non_NA_gt], xlim=c(0,100),
     main="Histogram of gq in non-NA genotypes", xlab="gq")
abline(v=min.gq, lty=2, col="red")

# dp (when gt is not NA !)
hist(dp.mx[non_NA_gt],
     main="Histogram of dp in non-NA genotypes", xlab="dp")
abline(v=max.dp, lty=2, col="red")

hist(dp.mx[non_NA_gt & dp.mx < 100], xlim=c(0,100),
     main="Histogram of dp in non-NA genotypes (zoom to <100)", xlab="dp")
abline(v=min.dp, lty=2, col="red")

# Mean GQ and DP in non-NA genotypes
quantile(gq.mx[non_NA_gt], na.rm=TRUE)
quantile(dp.mx[non_NA_gt], na.rm=TRUE)

mean(gq.mx[non_NA_gt], na.rm=TRUE)
mean(dp.mx[non_NA_gt], na.rm=TRUE)

# Clean-up
rm(non_NA_gt, gq.mx, min.gq, min.dp, max.dp)

```

# Variant call rates after filtering

```{r}

# Call rates per variant
var_call_rate <- apply(gt_add.mx,1,call_rate.udf)

hist(var_call_rate, breaks=50, main="Call rate per variant (after filters)", xlab="Call rates")
abline(v=0.85, lty=2, col="red")

plot(var_call_rate, main="Call rate per variant (after filters)")
abline(h=0.85, lty=2, col="red")

# SKAT doesnt allow variant call rate <85%
sum(var_call_rate < 0.85)/length(var_call_rate)

# Clean-up
rm(var_call_rate)

```

# Sample call rates after filtering

```{r}

# Call rates per sample
case_call_rate <- apply(gt_add.mx,2,call_rate.udf)

hist(case_call_rate, xlim=c(0,1), main="Call rate per sample (after filters)", xlab="Call rates")
abline(v=0.5, lty=2, col="red")

plot(case_call_rate, ylim=c(0,1), main="Call rate per sample (after filters)")
abline(h=0.5, lty=2, col="red")

low_call_rate_indices <- which(case_call_rate < 0.5)
low_call_rate_values <- case_call_rate[low_call_rate_indices]
low_call_rate_samples <- substr(names(low_call_rate_values),1,3)

text(x=low_call_rate_indices,y=low_call_rate_values,
     labels=low_call_rate_samples,pos=c(1,3,1), cex=0.7)

low_call_rate_values

# Clean-up
rm(call_rate.udf,case_call_rate,
   low_call_rate_indices,low_call_rate_values,low_call_rate_samples)

```

# Check data

```{r}

# List objects
ls()

# Check sizes
dim(gt_add.mx)
dim(gt_num.mx)
dim(gt_chr.mx)
dim(dp.mx)
dim(alt_fraction.mx)

dim(variants.df)
dim(phenotypes.df)
table(phenotypes.df$cc, useNA = "always")

# Check consistency of rownames and colnames

sum(rownames(gt_add.mx) != rownames(gt_num.mx))
sum(rownames(gt_add.mx) != rownames(gt_chr.mx))
sum(rownames(gt_add.mx) != rownames(dp.mx))
sum(rownames(gt_add.mx) != rownames(alt_fraction.mx))

sum(colnames(gt_add.mx) != colnames(gt_num.mx))
sum(colnames(gt_add.mx) != colnames(gt_chr.mx))
sum(colnames(gt_add.mx) != colnames(dp.mx))
sum(colnames(gt_add.mx) != colnames(alt_fraction.mx))

sum(rownames(gt_add.mx) != rownames(variants.df))
sum(colnames(gt_add.mx) != rownames(phenotypes.df)) 

```

# Save results

```{r}

save.image(file.path(data_folder,"s01_filter_genotypes_by_dp_and_gq.RData"))

```

# Final section

```{r}

sessionInfo()
Sys.time()
gc()

```
