---
title: "Select samples"
author: "Alexey Larionov"
date: "25 October 2020"
output:
  html_document:
    toc: true
    toc_float: true
editor_options:
  chunk_output_type: console
---

# Summary

Remove  

- NFE samples and  
- WECARE samples failed sequencing and other filters:  
  + low concordance with GWAS (when GWAS available)  
  + carriers of BRCA1/2 and PSALB2  
  + duplicates for QC  
  
Exclude non-polymotphic sites appeared after the samples removal.  

<style>
pre{
  overflow-x: auto;
}
pre code{
  word-wrap: normal;
  white-space: pre;
}
</style>

```{r echo=F}

options(width=999)

```

# Start section

```{r}

# Time
Sys.time()

# Memory
gc()

# Clean up
rm(list=ls())
graphics.off()

# Options
options(stringsAsFactors = F)

# Working folders
base_folder <- "/rds/project/erf33/rds-erf33-medgen"
project_folder=file.path(base_folder,"users/alexey/wecare/reanalysis_wo_danish_2020")
data_folder <- file.path(project_folder,"data","s04_add_phenotypes")
scripts_folder <- file.path(project_folder,"scripts","s04_add_phenotypes")
setwd(scripts_folder)

# Required library
library(dplyr)

```

# Load data

```{r}

load(file.path(data_folder,"s01_add_phenotypes.RData"))

```

# Check data

Note that phenotypes are not syncronised with other data yet  

```{r}

ls()

dim(full_variants.df)

dim(full_dp.mx)
dim(full_gq.mx)
dim(full_gt.mx)
dim(full_ad.mx)

dim(phenotypes_with_danish.df)

```

# Sequenced wecare samples

Ony wecare samples are present in the phenotype table (no NFE samples in the table).  

Note that  

- Some of the samples failed sequencing and  
- Some of the sequencing results do not represent any sample (e.g. if some samples were accidentally mixed during library preparation)  

```{r}

# Make an index for successfully sequenced wecare samples
sequenced_wecare <- phenotypes_with_danish.df$wes_id %in% colnames(full_gt.mx)

# Count samples
sum(sequenced_wecare)
sum(!sequenced_wecare) # wecare failed sequencing

```

# Danish samples

Danish consents do not allow studying known breast cancer risk genes, so danish samples are excluded from this analysis  

```{r}

# Make an index for danish samples
danish <- phenotypes_with_danish.df$danish

# Count samples
sum(danish)
sum(!danish)

```

# Other filters

Samples to be removed:  

- low concordance with GWAs  
- duplicates added for QC  
- BRCA1/2 and PALB2 carriers  

Keep eigenvector outliers sugested previously because eigenvectors will be re-calculated  

```{r}

# Explore filters
phenotypes_with_danish.df %>%
  group_by(filter) %>%
  summarise(count=n()) %>%
  arrange(desc(count))

# Select samples passed filters
pass_other_filters <- phenotypes_with_danish.df$filter == "pass" |
        phenotypes_with_danish.df$filter == "eigenvectors_outlier"

# Count samples
sum(pass_other_filters)
sum(!pass_other_filters)

```

# Combined filter

```{r}

# Combined filter (to use on rows of the phenotypes data frame)
combined_filter <- sequenced_wecare & !danish & pass_other_filters
sum(combined_filter)
sum(!combined_filter)

# Combined filter names (to use on columns of the genotype matrices)
selected_samples_names <- phenotypes_with_danish.df[combined_filter,"wes_id"]
length(selected_samples_names)

# Clean-up
rm(sequenced_wecare,danish,pass_other_filters) # combined_filter?

```

# Explore sample call rates

```{r}

# Function to calculate call rate
call_rate.udf <- function(x){sum(!is.na(x))/length(x)}

# Call rates in all genotypes
call_rates_all <- apply(full_gt.mx,2,call_rate.udf)

plot(call_rates_all, main="WECARE and NFE before filters")
abline(h=0.95, lty=2, col="red")
abline(v=198.5, lty=2, col="red")

# Call rates in selected non-danish wecare
call_rates_selected <- apply(full_gt.mx[,selected_samples_names],2,call_rate.udf)

plot(call_rates_selected, main="Non-Danish WECARE passed filters")

text(which(call_rates_selected < 0.935),
     call_rates_selected[call_rates_selected < 0.935],
     names(call_rates_selected[call_rates_selected < 0.935]),
     pos=4)

abline(h=0.9, lty=2, col="red")

# Clean-up
rm(call_rate.udf,call_rates_all,call_rates_selected,selected_samples_names)

```

# Update dataset

```{r}

# Update phenotypes table
selected_phenotypes.df <- phenotypes_with_danish.df[combined_filter,]

dim(selected_phenotypes.df)
summary(as.factor(selected_phenotypes.df$cc))

selected_phenotypes.df %>%
  group_by(filter) %>%
  summarise(count=n()) %>%
  arrange(desc(count))

# Get names of samples from the updated phenotypes table
selected_samples_names <- selected_phenotypes.df$wes_id

# Update genotype matrices
selected_dp.mx <- full_dp.mx[,selected_samples_names]
selected_gq.mx <- full_gq.mx[,selected_samples_names]
selected_gt.mx <- full_gt.mx[,selected_samples_names]
selected_ad.mx <- full_ad.mx[,selected_samples_names]

dim(selected_dp.mx)
dim(selected_gq.mx)
dim(selected_gt.mx)
dim(selected_ad.mx)

# Clean-up
rm(combined_filter,selected_samples_names,
   phenotypes_with_danish.df,full_dp.mx,full_gq.mx,full_gt.mx,full_ad.mx)

```

# Remove non-polymorphic variant sites

After removal of NFE and failed/danish WECARE samples some variant sites are becoming non-polymorphic in the dataset.

```{r}

# Function to detect uniform vector
uniformity_check.udf <- function(x){
  if (all(is.na(x))){"All_NA"}
  else if (min(x,na.rm=T)==max(x,na.rm=T)){"Uniform"}
  else {"Non-Uniform"}}
    
uniformity_check <- apply(selected_gt.mx,1,uniformity_check.udf)
summary(as.factor(uniformity_check))
non_uniform_sites <- uniformity_check == "Non-Uniform"

# Update variants data frame
selected_variants.df <- full_variants.df[non_uniform_sites,]

# Update genotype matrices
selected_dp.mx <- selected_dp.mx[non_uniform_sites,]
selected_gq.mx <- selected_gq.mx[non_uniform_sites,]
selected_gt.mx <- selected_gt.mx[non_uniform_sites,]
selected_ad.mx <- selected_ad.mx[non_uniform_sites,]

# Clean-up
rm(uniformity_check.udf,uniformity_check,non_uniform_sites,full_variants.df)

```

# Check data

```{r}

ls()

# Check objects dimentions
dim(selected_phenotypes.df)
dim(selected_variants.df)
dim(selected_dp.mx)
dim(selected_gq.mx)
dim(selected_gt.mx)
dim(selected_ad.mx)

# Check that the tables and matrices are in sync
sum(selected_phenotypes.df$wes_id != colnames(selected_dp.mx))
sum(selected_phenotypes.df$wes_id != colnames(selected_gq.mx))
sum(selected_phenotypes.df$wes_id != colnames(selected_gt.mx))
sum(selected_phenotypes.df$wes_id != colnames(selected_ad.mx))

sum(selected_variants.df$ID != rownames(selected_dp.mx))
sum(selected_variants.df$ID != rownames(selected_gq.mx))
sum(selected_variants.df$ID != rownames(selected_gt.mx))
sum(selected_variants.df$ID != rownames(selected_ad.mx))

```

# Save results

```{r}

save.image(file.path(data_folder,"s02_select_samples.RData"))

```

# Final section

```{r}

sessionInfo()
Sys.time()
gc()

```
